### PR TITLE
Added descriptor_type_version to ToolVersion definition

### PR DESCRIPTION
**Pull request by patmagee**
## What
I would like to propose adding a new field: `descriptor_type_version` to the `ToolVersion` object in order communicate the specific language level a particular `descriptor_type` is written in. The field is a simple `map` in the tool version, where the keys correspond to on of the `descriptor_type`'s and the values are the actual underlying  version.

Adding a new property does add a level of redundancy (ie reporting the `descriptor_type` twice). However, it does maintain backwards compatibility with the current spec version

```json
{
  "descriptor_type": [
    "CWL",
    "WDL"
  ],
  "descriptor_type_version": {
    "CWL": "v1.0.2",
    "WDL": "1.0"
   }
}
``` 

## Rationale
In `WES` there is a required field when submitting a new run: `workflow_type_version`. Workflow execution engines may handle run submissions differently depending on the reported language level. For example, an engine may only support version `1.0` of WDL or `v1.0.2` of CWL. It may also use different workers depending on the language version specified by the user. 

It would be my hope that we could relax the requirement in `WES` to require the `worflow_type` and `workflow_type_version` when a `TRS` uri is used. Unfortuantely, since `TRS` does not report the underlying version, we do not have the needed features to programatically look up this information without parsing the descriptor. Parsing the descriptor to get the version is not always possible as some languages do not report the version directly in their descriptor  

Additionally, when building applications that interact with tools/tool versions, it would be helpful to have this information available to build version specific elements in UI's or other tooling

### CODE REVIEW

File: openapi/openapi.yaml
Comment (denis-yuen) [inline openapi/openapi.yaml:633]: Should this be of `type` array?
My understanding is that CWL workflows can have a mixture of tools of different cwl versions.
https://github.com/common-workflow-language/cwltool/issues/1321

File: openapi/openapi.yaml
Comment (coverbeck) [inline openapi/openapi.yaml:641]: I don't know Nextflow, but looking at a random NF-Core workflow registered on Dockstore, it [specifies a minimum Nextflow version](https://github.com/nf-core/rnaseq/blob/1.4.2/nextflow.config#L166). Then I found the Nextflow [docs on it](https://www.nextflow.io/docs/latest/config.html?highlight=nextflowversion#nextflow-version). 

I'm personally more used to the [NPM semantic versioning](https://docs.npmjs.com/about-semantic-versioning), 

In any case, it seems like this case should be covered one way or another.

File: openapi/openapi.yaml
Comment (patmagee) [reply openapi/openapi.yaml:633]: @denis-yuen that makes sense. I will switch this over to a an array on Monday.

One thing I did consider, was changing the existing `descriptor_type` to contain this information since it is essentially being duplicated. But I figured that would be a harder change to get into the spec then a new field

File: openapi/openapi.yaml
Comment (denis-yuen) [reply openapi/openapi.yaml:641]: Hmmm. This seems a little different. 
The version of the Nextflow jar that a workflow is compatible with is different from the language version. 
Nextflow's equivalent of WDL draft-3, 1.0, etc. is DSL2 https://www.nextflow.io/docs/latest/dsl2.html

This does mean I would recommend adjusting the example though.

Comment (uniqueg) [reply openapi/openapi.yaml]: Would be good to hear something about the relationship between workflow languages/types, workflow language/type versions and required workflow engine versions from the developers of workflow languages for which only a single engine exists. Particularly for Snakemake, I don't think there is such a thing as a workflow language version (but I may be wrong). That would help us to better map workflows to WES instances and increase the chance that a workflow will actually be processed successfully.

I suppose the most interesting question would probably be:
* Do Nextflow/Snakemake workflows require specific workflow engine versions?
* Do Nextflow/Snakemake workflows require specific workflow language versions? Is there a concept of language versions at all in Snakemake?
* Are Nextflow and Snakemake workflow engines strictly backwards compatible in their ability to process older workflows / workflow languages?

Perhaps @johanneskoester and @pditommaso would like to comment?

File: openapi/openapi.yaml
Comment (uniqueg) [inline openapi/openapi.yaml:635]: `[...] actual language versions` (plural)

**Review by uniqueg — DISMISSED** [review]
Apart from input by workflow (language/engine) devs, there are a couple of minor wording issues. Otherwise it looks good to me. Very useful.

**Review by denis-yuen — APPROVED** [review]
think this is fine from my end, but worth pinging notifications

**Comment by denis-yuen** [comment]
Note, this handles one part of https://github.com/ga4gh/tool-registry-service-schemas/issues/210

**Comment by denis-yuen** [comment]
Note versioning convention, looks like we use the version number as-is without modification from the language itself

**Comment by denis-yuen** [comment]
Discussion from GA4GH call: what would break if we only reported the version from the root descriptor?

**Comment by denis-yuen** [comment]
note: there may be a PR in WES that deals with versioning https://github.com/ga4gh/workflow-execution-service-schemas/issues/181

**Comment by vsmalladi** [comment]
I think as long as you use the version from the tool itself there should be no issues. 

Not all workflows and tools will have backwards compatibility so there might be a minimum version if met would confer that a tool could work. But if we match exact I think this should be fine.

**Review by uniqueg — COMMENTED** [review]
The PR looks good to me and addresses an important use case. I do feel though that here may be a good opportunity to also address workflow types for which versions are not available, e.g., Snakemake (where workflow engine versions are used instead). There is a related PR in WES (https://github.com/ga4gh/workflow-execution-service-schemas/pull/182), and I would ideally like to see how negotiations to use a specific WES for a given workflow could/would play out for the different workflow types based on TRS URIs and the changes proposed here and in https://github.com/ga4gh/workflow-execution-service-schemas/pull/182.

Of course, this could also be addressed in future PRs, but there may be the risk that implementing these use cases might then require rolling back some changes or introducing breaking changes. However, if others feel that this risk is small/acceptable, from my side, the PR can be merged as is.

**Comment by denis-yuen** [comment]
I think the risk of rolling back changes is somewhat mitigated since this is marked as an optional field and not a required field. Pinging @patmagee for a final call, we should probably decide whether this is in or out for 2.0.1 in the next couple of days.

File: openapi/openapi.yaml
Comment (coverbeck) [inline openapi/openapi.yaml:636]: Nit: rest of the file uses `e.g.` instead of `ex`.

File: openapi/openapi.yaml
Comment (coverbeck) [inline openapi/openapi.yaml:641]: Even though it's obvious because it's an array, an example with two versions would be nice.

File: openapi/openapi.yaml
Comment (coverbeck) [inline openapi/openapi.yaml:736]: > For example, tools defined in CWL may have a version of `v1.0.2`

This reads a little prescriptive to me, e.g., they may not have a version v1.0.0. I suggest "could" instead of "may"

**Review by coverbeck — APPROVED** [review]
Some nit-picky suggestions.

It seems like a next step could be to surface this at the file level as well, although WES may not need it. And that gets trickier, because the file FileWrapper also describes non-workflow descriptors. But it feels incomplete, that the version says it has versions x and y, but there is no way to know which descriptors are which versions.

Just thinking out loud, and this shouldn't block the PR.

**Pull request merged**
